SNDINFO

```diff
@@ -559,7 +559,7 @@ $limit FAREXPL 4
 $rolloff  Explosion 1200 1800
 $rolloff  LightExplosion 1200 1800
 
-$random barrel/flameon		{ FLAMEON1 FLAMEON2 FLAMEON3 }	
+$random barrel/flameon		{ FLAMEON1 /*FLAMEON2 FLAMEON3*/ }	
```

Comment:
But I don't understand this change, could you please explain it to me?

Reply:
think it was just done as some minor temporary cleanup, theres probably a lot of stuff that points at missing sound files, and the worst part is, GZDoom doesnt really tell ya when there even are SNDINFO stuff that fails to load right, it just keeps on trucking and ignoring it

Reply:
I commented it out because the SNDINFO entries below are just duplicate mentions of FLAMEON1, but there might be additional FLAMEON sounds in the future.

Reply:
Oh, I see. GZDoom code is actually... hmm, I won't say it because I don't want to hurt nobody's feelings. But I've been diving into it in the last three days. I'd like to fork GZDoom to try to implement some basic optimizations that I do not understand how they haven't been done yet. I'll probably do it in a couple of weeks when I finish some current projects. My intention is to try to merge back my code in GZDoom if I manage to achieve some actual improvements (I've not coded in C++ since I was studying at the university, so I'm probably not the best C++ developer, but I've been working at the industry a lot of years and I can DEFINITELY identify ugly code in nearly any language)

If my pull requests are rejected I'm probably starting my own GZDoom fork to test some deeper optimizations like moving a lot of calculations the CPU should not be doing to de CPU, move some other unnecessary CPU calculations to GLSL shaders where they belong (like the sprite upscaling with xBRZ or whatever algorithm you choose), create a singleton Thread Pool and possibly a Fibers/Coroutines Pool to disengage and parallelize a lot of tasks from the main game loop (which seems to be untouched... since... like 2004?), simplify the use of shaders specially post-processing shaders and add some cool ones by default like Contrast Adaptative Sharpening or Contrast Limited Sharpening, analyze how it's possible that GZDoom uses more RAM than a modded Skyrim with 16K textures and more than 200GB of mods (maybe it's necessary to implement/improve its Virtual file system and introduce some kind of lazy loading because for the amount of RAM it uses, it's possible that it is loading in memory all mods/assets, even unused ones in the specific map you are playing), paralellize and improve the AABB tree and BSP rendering or replace them entirely with a more modern approach of GPU-based frustrum & occlussion culling... and I could go on because the list of ugly stuff I've seen in this three days is vast. Oh, and my ultimate goals, which are implementing LuaJIT as an alternative to ZScript (I simply can't understand why they choose to design their own language instead of using a faster standarized one like Lua or AngelScript) and possibly Ultralight to allow (optionally) writing the UI in HTML/JS/CSS. I probably don't achieve half of this, or at least not without some months of hard testing and development

Anyway, sorry for the TL;DR, I'm telling you this because I'll write this specific issue in my TODO list to check if it has a more or less simple fix. And as modders, if there is any feature/fix you'd like to be implemented, feel free to ask. I'm not doing you a favor, more like the opposite, you are doing me a favor asking for features or fixes, because my goal is to improve my C++ & 3D rendering skills to start working on my own game in a year or two. So I see this as an opportunity to both improve my skills and give back something to the community that has provided me SO many hours of entertainment

By the way, this branch seems to be working fine. I'll keep playing in a couple of ours. Only issue I have found, but completely unrelated to this mod or this PR, is that the upscaled sprites I'm working on make GZDoom use 14GB of RAM. As I said, it's insane

Reply:
> analyze how it's possible that GZDoom uses more RAM than a modded Skyrim with 16K textures and more than 200GB of mods (maybe it's necessary to implement/improve its Virtual file system and introduce some kind of lazy loading because for the amount of RAM it uses, it's possible that it is loading in memory all mods/assets, even unused ones in the specific map you are playing),

AFAIK GZDoom does just load assets into memory as needed, but it has no streaming so they just stay there, with a few drop off points like level changes.

> paralellize and improve the AABB tree and BSP rendering or replace them entirely with a more modern approach of GPU-based frustrum & occlussion culling...

dpJudas already has a WIP branch that does something like that I think, after he lost the original 2018 version due to only hoarding it for himself. And [Helion](https://www.doomworld.com/forum/topic/132153-helion-c-0916-0923-goodbye-bsp-tree-rendering/) is basically built around what you seem to be describing.

> Oh, and my ultimate goals, which are implementing LuaJIT as an alternative to ZScript (I simply can't understand why they choose to design their own language instead of using a faster standarized one like Lua or AngelScript) and possibly Ultralight to allow (optionally) writing the UI in HTML/JS/CSS.

Yeah I think that will just get instantly rejected even if you somehow implement it flawlessly, and honestly it sounds super unnecessary. To add a THIRD scripting language, there's already DECORATE (Deprecated) and ZScript.  As for why not Lua, well, I think Randi did actually consider it at some point, and Sonic Robo Blast 2 does just flat out use Lua. But ZScript was most likely picked for a similar reason to why Godot prefers GDScript over C#, because it's more or less directly tied into the engine instead of being an external interface. And writing UI code in fucking web languages for Doom just sounds comical.

> And as modders, if there is any feature/fix you'd like to be implemented, feel free to ask. I'm not doing you a favor, more like the opposite, you are doing me a favor asking for features or fixes, because my goal is to improve my C++ & 3D rendering skills to start working on my own game in a year or two.

Hmm, well. I guess this would be the list of stuff GZDoom needs in order of importance.

Implementing a native navigation mesh and pathfinding solution. There have been many attempts to make NPCs that don't stupidly bash their skulls on walls or get stumped by a bucket-shaped wall, but most either flat out failed, are very janky and likely don't even work most of the time, are slow due to being mod implementations etc. Having a native navigation mesh and pathfinding system (That is of course moddable through and exposed ZScript) would be fucking great, and would be faster than any hacky mod implementation or a pure ZScript solution. [Recast seems like a good pick.](https://github.com/recastnavigation/recastnavigation) [And is what one of the more promising mod pathfinding attempts used too, but it was pretty buggy from my experience, and the person behind it stopped updating it since it already accomplishes the simple pathfinding he needed for his game.](https://github.com/disasteroftheuniverse/zdoom-pathfinding) It also relied on an ancient version of Recast, since he used some Javascript (I think) port of it, due to being what he's familiar with mostly.

I'd be nice to finally have 3D polyobjects in GZDoom [using K8Vavoom's implementation, which already has a spec for how it works](https://repo.or.cz/k8vavoom.git/tree/HEAD:/specs/3dpobj), instead of using dumb ZScript hacks to make semi-functional horizontally moving platforms that require different actual models for every single different platform.

Being able to mark sectors as static (Not needing to be updated constantly) would also be useful, especially for huge maps where the BSP tanks the framerate.

[Having flat decals instead of having to use 3D would really help Project Brutality and basically every other mod, probably even more so if damage and such can be attached to them like in K8Vavoom](https://repo.or.cz/k8vavoom.git/blob/HEAD:/specs/k8vavoom_decals.txt). K8Vavoom even has decals that are clipped based on texture alpha, so something like a blood splatter on a bar mid texture or a 3D floor transparent grating will be alpha clipped. Instead of decals just not being splatted on semi-transparent surfaces at all.

There's more that I and I'm sure a lot of other people would like to see. But these are the ones I can actually remember right now that are not getting added any time soon. If you were to do this like last year the list would be a lot larger, but it seems that strangely enough GZDoom is actually progressing. RaveYard and dpJudas are working on proper 3D lighting and lightmaps (Which K8Vavoom has had for years), but of course the system requirements are higher and AFAIK there won't be any[ projected sprite shadows](https://www.youtube.com/watch?v=4O7RT65tXB0). MajorCooke is working on a native scriptable particle, which means that we will finally have a scriptable actor-esque thinker with a world position and rendering. For making simple effects without using the bloated and slow as balls AActor class.

Reply:
If you do choose to get into the CPU/GPU optimization side of GZDoom, this would be a godsend. Not going to throw any shade on the GZDoom devs as they've been doing what they can with the time they have. But the things you're looking to do sound like some really in-depth rewriting and rethinking of code that's been there for decades.  Fact is, when stuff has been in there that long, folks don't want to mess with it for fear of breaking stuff. But, that's not to say that having a fresh set of eyes looking at it and trying new approaches is a bad thing at all. If it ends up working, improving performance, and retaining compatability, then I doubt anyone would have an issue with it.
For myself, I work on a mod/mapset that in certain areas absolutely brings GZDoom to its knees. So if it can improve, I'm all for it. Those camera textures probably need to be looked at. That said, if modders don't push things further and find where the limits are, things never progress, so in a way, I'm kind of happy to be making something that is so heavy.